rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -237,50 +243,56 @@ pub fn check_for_pre_action(pre_action_info: &mut PreAction, items: &mut Vec<Ite
                 continue;
             }
             if num_pa_attrs_on_fn > 1 {
-                abort!(item, "More than one pre_action attribute found.");
+                return Err(
+                    syn::Error::new(f.span(), "More than one pre_action attribute found.")
+                        .to_compile_error(),
+                );
             }
             num_pre_action_fns += 1;
-            if num_pre_action_fns > 1 {
-                abort!(item, "No more than 1 pre_action fn permitted.");
-            }
+
             let pre_action_attr = pre_action_attrs.get(0).unwrap().clone();
+            sp = Some(pre_action_attr.clone());
 
             pre_action_info.fn_name = Some(f.sig.ident.clone());
 
-            let attr_args_ts = match pre_action_attr.meta.clone() {
-                Meta::List(args) => args.tokens,
-                Meta::Path(args) => abort!(
-                    args,
-                    "Invalid pre_action attributes: expected list; got path."
-                ),
+            pre_action_info.exclude = match pre_action_attr.meta.clone() {
+                Meta::List(args) => {
+                    let attr_args = NestedMeta::parse_meta_list(args.tokens.clone())
+                        .unwrap_or_else(|err| {
+                            abort!(
+                                args.tokens,
+                                format!("Invalid pre_action arguments: {}", err.to_string())
+                            )
+                        });
+
+                    let options: PreActionOptions = PreActionOptions::from_list(&attr_args.clone())
+                        .unwrap_or_else(|err| {
+                            abort!(
+                                args.tokens,
+                                format!("Invalid pre_action arguments: {}", err.to_string())
+                            );
+                        });
+
+                    options.exclude
+                }
+                Meta::Path(_) => PathList::new::<syn::Path>(vec![]),
```

Comment:
While testing, I found that this branch was erroring when I intended it to work, just without specifying exclude; so this is the fix.